### ISSUE TITLE
creating entity IDs from squuids

### ISSUE THREAD

**Issue by thedavidmeister** [tonsky/datascript#168]
Hi, I'm trying to figure out how to keep entity IDs created locally by datascript in a central database.

The IDs in the central database need to be more universally unique, to allow more than one client access to the data at a time.

I've written a bit of code to generate a squuid for each entity ID as it is sent to the remote server and make sure that squuids and entity IDs are matched consistently across multiple transactions/datoms.

I'm now not sure what the best way to get squuids back into entity IDs on a fresh page load (or injected into an existing database) would be.

I have a few ideas but I'm not totally happy with any of them:

- Use SQUUIDS as entity IDs in datascript -> currently unsupported, throws an error because datascript expects integer entity IDs
- Increment a counter for each new squuid hitting the browser from the database, use the counter as the entity ID -> there is no guarantee that squuids will be sent to the browser in the order they were created, making the order of the entity IDs different to the order of the squuids (which is bad if you rely on the order of entity IDs)
- Use the time component of the SQUUID directly as the entity ID -> this results in entity IDs that are larger than the start of transaction IDs, leading to at least one pretty bad bug in `advance-max-eid` where the database's `max-eid` is not updated. Even if we subtract a big number from the time before using it, we can still easily get collisions as the time component is only accurate to the second

**Comment by tonsky** [comment]
Practically, I’d probably suggest you to go with third option. Yes, max-eid is not advanced, but you can see it as a plus: you’ve taken id management into your own hands, so DataScript algorithm steps away. You can actually mix both, continue using DS-generated ids at the same time, because they won’t overlap in any foreseeable future.

There’s a fourth option, actually, which we kind of use right now in Cognician when syncing between Datomic and DataScript. You can’t tell Datomic which exact entity id you want (unlike DS, which is unfortunate). So we’re not relying on entity ids at all: each entity also has an `:.../uuid` attribute (`:user/uuid`, `:resource/uuid`, `:session/uuid` etc). When syncing, we calculate dependencies based on uuids, not entity ids. In fact, we don’t care about entity ids and if they match between Datomic and DataScript. We go as far as don’t send entity ids over the wire at all. We rely on default allocator in both. But we track uuids and make sure dependencies are tracked right in each case. It’s a little more work, and I’d certainly prefer not doing it if that was possible (e.g. if Datomic’d allowed for manually assigning entity ids), but it works ok in practice.

**Comment by thedavidmeister** [comment]
@tonsky thanks for the help! i'm not 100% sure how your option 4 is different pragmatically to my option 2? (which is what I'm trying at the moment, the default allocator works the way that I described my counter works, i think).

The problem I'm facing (that I think you would face too) is that I'm trying to do simple chronological sorts of elements in the UI by calling `datoms` and using the `:avet` index. This totally breaks down pretty quickly when dealing with removing the original eids in the backend storage, and as soon as you're expecting some other client to send you new datoms/entities, they may need to insert an eid halfway through your eids after latency, etc...

The chronological sort seems like a pretty common use case to me because that basically means that as users create things in the UI from top to bottom, things will continue to be laid out in the same order on the next page refresh. However, this might well end up being too naive and I'll need to add timestamps or weights as attributes to entities later.

Anyway, I have a few questions to help me get my head around the problem:

- Do you know of any other issues that going past the `tx0` limit might cause, other than the auto increment?
- I saw that some performance issues might arise (https://github.com/tonsky/datascript/issues/56), any idea how much of a concern this is in practise?

**Comment by thedavidmeister** [comment]
also, is there any way to hook into the way that DS chooses new entity IDs, either through `next-eid` or messing with the temp IDs through `listen!` or similar?

**Comment by tonsky** [comment]
> Do you know of any other issues that going past the tx0 limit might cause, other than the auto increment?

Nothing else

> I saw that some performance issues might arise (#56), any idea how much of a concern this is in practise?

I guessed that, no idea how serious it is.

> also, is there any way to hook into the way that DS chooses new entity IDs, either through next-eid or messing with the temp IDs through listen! or similar?

No, but you can always put a specific id you want into transaction

**Comment by wbrown-lg** [comment]
@thedavidmeister An alternative is to hash the squuid into a `2^53` bit number, but depending on how many entities you need to be tracking, you risk collision with a 53 bit resolution of the hash.  The nice thing about this is that the datoms are universally consistent, and can be cached and transmitted over the wire.  You can literally concatenate arrays of Datoms that meet your criteria and build a Datascript database out of them.

**Comment by thedavidmeister** [comment]
@wbrown-lg yeah i ended up doing a bit of "translation" of eids on datoms between client and backend db, just using a millisecond timestamp atm as for my use-case it currently only needs to be unique per-user.
